internal/cmd/oohelperd/internal/nwcth/measure.go

```diff
@@ -0,0 +1,281 @@
+		return nil, errors.New("no valid IP address to measure")
+	}
+
+	wg := new(sync.WaitGroup)
+	out := make(chan *MeasureEndpointResult, len(enpnts))
+	for _, endpoint := range enpnts {
+		wg.Add(1)
+		go MeasureEndpoint(ctx, creq, URL, endpoint, wg, out)
+	}
+	wg.Wait()
+	close(out)
+
+	h3Reqs := []*CtrlRequest{}
+	redirectedReqs := []*CtrlRequest{}
+	for m := range out {
+		urlMeasurement.Endpoints = append(urlMeasurement.Endpoints, m.CtrlEndpoint)
+		if m.httpRedirect != nil {
+			req := &CtrlRequest{HTTPCookieJar: m.httpRedirect.jar, HTTPRequest: m.httpRedirect.location, HTTPRequestHeaders: m.httpRedirect.httpRedirectReq.Header}
+			redirectedReqs = append(redirectedReqs, req)
+		}
+		if m.h3Location != "" {
+			req := &CtrlRequest{HTTPRequest: m.h3Location}
+			h3Reqs = append(h3Reqs, req)
+		}
+	}
+	return &MeasureURLResult{CtrlURLMeasurement: urlMeasurement, h3Reqs: h3Reqs, redirectedReqs: redirectedReqs}, nil
+}
+
+func MeasureEndpoint(ctx context.Context, creq *CtrlRequest, URL *url.URL, endpoint string, wg *sync.WaitGroup, out chan *MeasureEndpointResult) {
+	defer wg.Done()
+	endpointResult := &MeasureEndpointResult{}
+	measureFactory[URL.Scheme](ctx, creq, endpoint, wg, endpointResult)
+	out <- endpointResult
+}
+
+var measureFactory = map[string]func(ctx context.Context, creq *CtrlRequest, endpoint string, wg *sync.WaitGroup, result *MeasureEndpointResult){
+	"http":  measureHTTP,
+	"https": measureHTTP,
+	"h3":    measureH3,
+	"h3-29": measureH3,
+}
+
+func measureHTTP(
+	ctx context.Context,
+	creq *CtrlRequest,
+	endpoint string,
+	wg *sync.WaitGroup,
+	result *MeasureEndpointResult,
+) {
+	URL, _ := url.Parse(creq.HTTPRequest)
+	httpMeasurement := CtrlHTTPMeasurement{Endpoint: endpoint, Protocol: URL.Scheme}
+	var conn net.Conn
+	conn, httpMeasurement.TCPConnect = TCPDo(ctx, &TCPConfig{
+		Endpoint: endpoint,
+	})
+	if conn == nil {
+		return
+	}
+	defer conn.Close()
+	var transport http.RoundTripper
+	switch URL.Scheme {
+	case "http":
+		transport = nwebconnectivity.GetSingleTransport(conn, nil)
+	case "https":
+		var tlsconn *tls.Conn
+		cfg := &tls.Config{ServerName: URL.Hostname()}
+		tlsconn, httpMeasurement.TLSHandshake = TLSDo(ctx, &TLSConfig{
+			Conn:     conn,
+			Endpoint: endpoint,
+			Cfg:      cfg,
+		})
+		if tlsconn == nil {
+			return
+		}
+		transport = nwebconnectivity.GetSingleTransport(tlsconn, cfg)
+	}
+	// perform the HTTP request: this provides us with the HTTP request result and info about HTTP redirection
+	httpMeasurement.HTTPRequest, result.httpRedirect = HTTPDo(ctx, &HTTPConfig{
+		Jar:       creq.HTTPCookieJar,
+		Headers:   creq.HTTPRequestHeaders,
+		Transport: transport,
+		URL:       URL,
+	})
+	// find out of the host also supports h3 support, which is announced in the Alt-Svc Header
+	h3Support := discoverH3Server(httpMeasurement.HTTPRequest, URL)
+	if h3Support != "" {
+		quicURL, _ := url.Parse(URL.String())
+		quicURL.Scheme = h3Support
+		result.h3Location = quicURL.String()
+	}
```

Comment:
I think you could refactor this snippet of code so that the helper function gives you the correct H3 URL to use or an error. (This comment takes into account my previous comment that `discoverH3Server` does not parse many compliant `Alt-Svc` headers and should therefore be extended.)